python/sandbox/das_gate/dasgate.py

```diff
@@ -0,0 +1,94 @@
+from hyperon import *
+from hyperon.ext import register_atoms
+
+import os
+import re
+import json
+
+from hyperon_das import DasAPI
+from hyperon_das.das import QueryOutputFormat as Format
+from hyperon_das.pattern_matcher import (
+    Link,
+    Node,
+    Variable,
+    And,
+    Or,
+    Not,
+    PatternMatchingAnswer,
+)
+
+class DASpace(AbstractSpace):
+
+    def __init__(self, unwrap=True):
+        super().__init__()
+        self.das = DasAPI('hash_table')
+        self.unwrap = unwrap
+
+    def _atom2dict(self, atom):
+        if isinstance(atom, ExpressionAtom):
+            return {
+                "type": "Expression",
+                "targets": [self._atom2dict(ch) for ch in atom.get_children()]
+            }
+        else:
+            return {
+                "type": "Symbol",
+                "name": repr(atom)
+            }
+
+    def _atom2query(self, atom):
+        if isinstance(atom, ExpressionAtom):
+            targets = atom.get_children()
+            if isinstance(targets[0], SymbolAtom) and targets[0].get_name() == ',':
+                return And([self._atom2query(ch) for ch in targets[1:]])
+            return Link("Expression", ordered=True,
+                targets=[self._atom2query(ch) for ch in targets])
+        else:
+            if isinstance(atom, VariableAtom):
+                return Variable(repr(atom))
+            else:
+                return Node("Symbol", repr(atom))
+
+    def _handle2atom(self, h):
+        try:
+            return S(self.das.get_node_name(h))
+        except Exception as e:
+            return E(*[self._handle2atom(ch) for ch in self.das.get_link_targets(h)])
+
+    def query(self, query_atom):
+        query = self._atom2query(query_atom)
+        answer = PatternMatchingAnswer()
+        new_bindings_set = BindingsSet.empty()
+        if not query.matched(self.das.db, answer):
+            return new_bindings_set
+        for a in answer.assignments:
+            bindings = Bindings()
+            for var, val in a.mapping.items():
+                # remove '$', because it is automatically added
+                bindings.add_var_binding(V(var[1:]), self._handle2atom(val))
+            new_bindings_set.push(bindings)
+        return new_bindings_set
+
+    def add(self, atom):
+        dc = self._atom2dict(atom)
+        if "name" in dc:
+            self.das.add_node(dc)
+        else:
+            self.das.add_link(dc)
+
+    #def remove(self, atom):
```

Comment:
Without `remove` and `replace` implementations, an exception will be raised if they're called.

Do we want to modify the Space contract so that a Space can gracefully decline to implement these functions in the same way a Space can decline to implement `atom_count` and `atoms_iter`?  My feeling is "no", but you have a more coherent conceptual vision for MeTTa than I do.

It seems to me that `remove` and `replace` are different from `atom_count` and `atoms_iter` in that `remove` and `replace` affect the Space.  Even in a space where it's impossible to "forget" (e.g. an append-only journal) it's still possible to indicate that an atom is disused or has been superseded.

Reply:
> Without remove and replace implementations, an exception will be raised if they're called.

This is perfectly ok. 

> Do we want to modify the Space contract so that a Space can gracefully decline to implement these functions

What behavior will be in this case for `remove`, when it is not implemented? I believe it should be exception. A space may not implement some required functionality, but attempts to use the missing functionality should be prevented, and the user should be informed that there is no such functionality and it cannot be used.

> in the same way a Space can decline to implement atom_count and atoms_iter?

Do you mean that if `atoms_iter` is not implemented, and `get-atoms` is executed in MeTTa, a MeTTa `Error`-expression is returned, while if `remove-atom` is executed the exception is raised, which is not turned into `Error`, but the whole MeTTa runner stops? I'm ok with this difference.

Reply:
>  if remove-atom is executed the exception is raised, which is not turned into Error, but the whole MeTTa runner stops? I'm ok with this difference.

Does it really work by this way? I believe the exception should be wrapped into `Error` as well.